### PR TITLE
Actor API Proposal

### PR DESCRIPTION
**Pull request by Theomat**
Note that first of all `actor.py` will change based on #40, as such this should not be merged until #40 has been merged then this should be adapted.

This is a proposal of the **RogueNet** API, so everything is up for debate.

You create an `Actor` for the neural net part of your agent.
First issue, the name is originally `Agent`, I am not too fond of that name since for me it entails not only the policy but the features as well hence why I changed the name to `Actor`.

The base class should take everything as a parameter and do the base logic. This enables in the long run easy sharing of embeddings or heads for centralized multi agent training I believe.
Then there is the `AutoActor` that uses the creator classes to automatically build the necessary embeddings and action heads.
The value head is part of the `auxiliary_heads` now which is just the general case, and `head_creator` has a method to create a value head.

Why creators? There might be a lot of specific logic depending on the space (either `ObsSpace` or `ActionSpace`) which should live in a separate file from the `Actor`, furthermore this enables users to use only part of the **RogueNet** framework to combine with their own, rather than include everything in a single class.

Take note of the `backbone_creator` which correctly leveraged should be enough to support a wide range of architectures out of the box.

### CODE REVIEW

File: rogue_net/rogue_net/head_creator.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from typing import Dict
+from torch import nn
+
+from entity_gym.environment import ActionSpace, CategoricalActionSpace
+
+
+def create_head_for(space: ActionSpace) -> nn.Module:
```
Comment (cswinter) [inline rogue_net/rogue_net/head_creator.py:7]: This should probably take `d_model` as an input?

**Review by cswinter — COMMENTED** [review]
This looks like a good refactor! I do expect a lot of this will change, e.g. the way we define action spaces will likely become a lot more complex, but it makes the most sense to adapt the interface once we know what we actually need.

I would do the entire move in one go, including https://github.com/entity-neural-network/incubator/issues/43 and deleting the files in enn_ppo, that way the code remains tested and git knows this is a rename which will make merges easier. Don't worry about waiting on https://github.com/entity-neural-network/incubator/pull/40 with this, I'll have to do a good amount of refactoring on the interface anyway and it might still a while until it's ready.

File: rogue_net/rogue_net/head_creator.py (hunk 1 above)
Comment (Theomat) [reply rogue_net/rogue_net/head_creator.py]: This has been changed

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -24,21 +24,23 @@ jobs:
...
-          key: ${{ hashFiles('requirements-dev.txt') }}-${{ hashFiles('entity_gym/setup.py') }}-${{ hashFiles('enn_ppo/setup.py') }}
+          key: ${{ hashFiles('requirements-dev.txt') }}-${{ hashFiles('entity_gym/setup.py') }}-${{ hashFiles('enn_ppo/setup.py') }}-${{ hashFiles('rogue_net/setup.py') }}
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
           pip install torch==1.10
           pip install -r requirements-dev.txt
           pip install -e entity_gym
           pip install -e enn_ppo
+          pip install -e rogue_net
```
Comment (cswinter) [inline .github/workflows/tests.yaml:35]: I would reorder this since `enn_ppo` depends on `rogue_net`.

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -24,21 +24,23 @@ jobs:
...
       - name: Black code formatting
         run: |
           black --check --diff .
       - name: MyPy
         run: |
           mypy entity_gym
           mypy enn_ppo
+          mypy rogue_net
```
Comment (cswinter) [inline .github/workflows/tests.yaml:43]: Same here.

**Review by cswinter — APPROVED** [review]
I have one nit but otherwise LGTM.

**Pull request merged**